File: 3213-count-subarrays-where-max-element-appears-at-least-k-times/3213-count-subarrays-where-max-element-appears-at-least-k-times.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var countSubarrays = function(nums, k) {
    const maxElement = Math.max(...nums);
    let result = 0;
    let end = 0;
    let maxCount = 0;

    for (let start = 0; start < nums.length; start++) {
        if (nums[start] === maxElement) maxCount++;

        while (maxCount === k) {
            result += nums.length - start;
            if (nums[end] === maxElement) maxCount--;
            end++;
        }
    }

    return result
};

File: 3213-count-subarrays-where-max-element-appears-at-least-k-times/3213-count-subarrays-where-max-element-appears-at-least-k-times.py
class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        maxElement = max(nums)
        result = 0
        end = maxFreq = 0

        for start in range(len(nums)):
            if nums[start] == maxElement:
                maxFreq += 1
            while maxFreq == k:
                result += len(nums) - start
                if nums[end] == maxElement:
                    maxFreq -= 1
                end += 1
                
        return result